### PR TITLE
Add better support for location ID filters

### PR DESCRIPTION
**Pull request by czue**
http://manage.dimagi.com/default.asp?186317 :blowfish: / :blowfish: 

first many commits are refactoring. actual change is added in https://github.com/dimagi/commcare-hq/commit/21085f4472c677303c51e62fd1c1daacab27d29b, docs in https://github.com/dimagi/commcare-hq/commit/6e6d79ec17726e491f376f8f3a35f73079e72092

This should also allow additional extensions (e.g. users, groups, etc.) to be dropped in super easily. I was thinking that someone more familiar with elasticsearch might take the lead on that.

people who might be interested: @esoergel @dannyroberts @sheelio @nickpell @NoahCarnahan 

cc @calellowitz

### CODE REVIEW

File: corehq/apps/userreports/views.py [hunk 1]
```diff
@@ -55,8 +53,10 @@
     get_datasource_config,
     get_report_config,
 )
+from corehq.apps.userreports.reports.filters.choice_providers import ChoiceQueryContext
+from corehq.apps.userreports.reports.filters.factory import FilterChoiceProviderFactory
```
Comment (dimagimon) [inline corehq/apps/userreports/views.py]: F401 'FilterChoiceProviderFactory' imported but unused

**Comment by sheelio** [comment]
@czue Does this filter by the chosen location and all of its descendants or just by the chosen location?

**Comment by czue** [comment]
just the chosen location for the time being.

**Comment by NoahCarnahan** [comment]
Nice! looks good to me. Ready to merge from my perspective, my `_asdict()` comment is non blocking, but I'll hold off on merging for a few minutes in case you want to respond.

**Comment by czue** [comment]
@NoahCarnahan updated

**Pull request merged**